Add Repo.load/2

wojtekmach opened the pull request:
Closes #1590.

I figured I'll open up a very early draft to collect some feedback and answer some questions before spending more time on this:

TODO:
- [x] Support string keys, e.g.: `Ecto.load(Post, %{"title" => "title 1"})`
- [x] Support atom keys, e.g.: `Ecto.load(Post, %{title: "title 1"})`
- [x] Support keyword, e.g.: `Ecto.load(Post, [title: "title 1"])`
- [x] Support tuple, e.g.: `Ecto.load(Post, {[:title], ["title 1"]})`
- [x] Better error handling (e.g. nicer error messages)?
- [x] Schema-less support? E.g.: `Ecto.load(%{title: :string}, %{title: "title 1"})`
- [x] Better unit tests - what else is worth testing?
- [x] Typespecs & docs

MAYBE:
- [ ] Support associations, e.g.: `Ecto.load(Post, %{title: "title 1", comments: [%{text: "text 1"}]`
- [x] ~~Support Ecto.load/3, e.g.: `Ecto.load(Post, ["title"], [["title 1"], ["title 2"]])` - worth keeping? Btw, I don't think association could be supported here.~~
- [x] Keep integration test?

Some of the TODO items are redundant per comments on #1590 but I thought it would be worth repeating them in order to potentially re-consider some features.

josevalim commented:
@wojtekmach I changed your issue description to move the things I don't believe we should support to a MAYBE section. I would prefer to keep this "low-level", so for example, no association handling.

wojtekmach commented:
Thanks for triaging @josevalim.

Come to think about it, implementation could be achieved by Changeset.cast + apply_changes too, but I guess since its gonna be low level its ok to call Schema.__load__ directly than go through changeset.

josevalim commented on lib/ecto.ex:
```diff
@@ -595,4 +595,25 @@ defmodule Ecto do
+  end
+
+  defp stringify_map_keys(map) do
+    Enum.into(map, %{}, fn
```
Since we will end-up traversing the map inside `Ecto.Schema.__load__`, I would change `Ecto.Schema.__load__` to also accept a keyword list and use that in the default implementation. The map implementation can call `Map.to_list` and then delegate to the keyword one.

wojtekmach replied:
Fixed in 40315f1 - not super happy with the impl., see inline comment

wojtekmach commented on lib/ecto/schema.ex:
```diff
@@ -1329,17 +1329,22 @@ defmodule Ecto.Schema do
-        case Map.fetch(map, Atom.to_string(field)) do
-          {:ok, value} -> Map.put(acc, field, load!(struct, field, type, value, loader))
-          :error -> acc
+        case Enum.find(data, fn {key, _value} -> key == field or key == Atom.to_string(field) end) do
```
needed to look for both atom as well as string keys. If it was just atom keys (`Keyword.fetch`), some integration & association tests were failing.

josevalim replied:
I am not a big fan. This code is used when loading data from the database too, so it should be as efficient as possible, and not concern about specific key types. I would convert any atom key to a string before hand. 

Another way to put this, we should avoid changing this code as much as possible.

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -1329,17 +1329,22 @@ defmodule Ecto.Schema do
 
-  defp do_load(struct, types, map, loader) when is_map(map) do
+  def __load__(schema, prefix, source, context, data, loader) when is_map(data) do
+    __load__(schema, prefix, source, context, Map.to_list(data), loader)
```
Sorry, I gave bad advice. We are not traversing the map, so keeping it as a map will be more efficient.

josevalim replied:
Glad to help! I think the previous version of this PR was closer to what we wanted, I have pushed you to the wrong direction, sorry about that.

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -1329,17 +1329,22 @@ defmodule Ecto.Schema do
+  end
+
+  @doc false
+  def do_load(struct, types, data, loader) when is_list(data) do
```
If we need to expose it, do it as `__load__` too.

wojtekmach commented on lib/ecto.ex:
```diff
@@ -595,4 +595,21 @@ defmodule Ecto do
     end
   end
+
+  @spec load(Ecto.Schema.t | map(), map() | Keyword.t | {list, list}) :: Ecto.Schema.t | map()
```
@josevalim I took a stab at defining a type spec for this. WDYT?

With latest changes we'd basically support 3 ways of passing data:

```elixir
Ecto.load(User, %{name: "joe"})
Ecto.load(User, [name: "joe"])
Ecto.load(User, {[:name], ["joe"]})
```

the 3rd one is very similar to @michalmuskala's original proposal to pass columns & values.

I finally realized that `Schema.__load__` was getting data primarly as `{fields, values}` (e.g. from DB), and `__load__` accepting a map was only used to fill in assoc/map fields.

One thing that is now changed is that `Ecto.load(User, %{"name" => "joe"})` is no longer supported out of the box (I could add key stringification to Ecto.load) and I think it's actually for the better - if it's supposed to be fairly low-level method then I think it's fine to assume that field names are atoms. (again, when `Ecto.__load__` is called by the code that grabbed data from DB, column names were already converted to atoms)

wojtekmach replied:
FWIW even though `{fields, values}` is being used under the hood, it might make sense to NOT expose it to keep the interface tighter (callers would just do `Enum.zip`)

wojtekmach replied:
btw, here's how it can be used end-to-end. Not sure it achieves Michal's original goal of making this stuff much easier, but perhaps it's a start:

```elixir
result = Ecto.Adapters.SQL.query!(HexWeb.Repo, "SELECT * FROM users", [])
columns = Enum.map(result.columns, &String.to_atom/1)
Enum.map(result.rows, fn row -> Ecto.load(HexWeb.User, Enum.zip(columns, row)) end)
#=> [%HexWeb.User{}, ...]
```

josevalim replied:
:+1: for the tuple approach. It is fine if we document it propelry.

josevalim commented on lib/ecto.ex:
```diff
@@ -595,4 +595,21 @@ defmodule Ecto do
+
+  @spec load(Ecto.Schema.t | map(), map() | Keyword.t | {list, list}) :: Ecto.Schema.t | map()
+  def load(schema_or_types, data) when is_map(data) do
+    load(schema_or_types, {Map.keys(data), Map.values(data)})
```
Let's pass the map as is since we already support maps in `__load__`?

michalmuskala replied:
The string key support in load is mainly for embeds that we get from database as a map with string keys (mongo also uses it as the basic loading mechanism, since we get maps back).

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -1329,7 +1329,12 @@ defmodule Ecto.Schema do
+  end
+
+  @doc false
+  def do_load(struct, types, map, loader) when is_map(map) do
```
`defp`? We should not export it.

wojtekmach replied:
yup, forgot to make it private again, will fix

josevalim replied:
Or even better, just make all of them a `__load__` and remove `def do_load`.

wojtekmach replied:
ok, I'll see how that would look like :+1:

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -1329,7 +1329,12 @@ defmodule Ecto.Schema do
   end
 
-  defp do_load(struct, types, map, loader) when is_map(map) do
+  def __load__(struct, types, {fields, values}, loader) when is_list(fields) and is_list(values) do
```
Need to add `@doc false` here.

josevalim commented:
@wojtekmach this looks great to me. There are some minor details but they are cosmetic (public/private functions). I think we can move to write the tests so we can finally merge this. :)

wojtekmach commented:
> I think we can move to write the tests so we can finally merge this. :)

I already added some tests, but I'm sure more could be added. I definitely need to add docs though!

josevalim commented:
> I already added some tests, but I'm sure more could be added. I definitely need to add docs though!

Gah, sorry again. I meant the docs.

josevalim commented:
I think we should support string keys otherwise folks will call String.to_atom when they need it and it will be "wrong". Also, `struct/2` supports strings. It is ok to add more clauses to `__load__` if necessary or to support both atom and string keys as long as we keep it fast.

michalmuskala commented:
@josevalim are you sue `struct/2` supports strings? From the docs:

>┃ # String keys are ignored
>┃ struct(User, %{"name" => "meg"})
>┃ #=> %User{name: "john"}

josevalim commented:
@michalmuskala yes, you are correct. I don't know what is happening with me...

wojtekmach commented:
@josevalim I added support for string keys in 675e1a6. Here's some context: (sorry if it's redundant, but helps me clear my thinking!)

on master, `__load__` handles passed `data` in two ways:

1. data comes as `[field|fields]`, `[value|values]`, and when `Map.fetch(types, field)` returns `:error` then we raise an error.
2. data comes as `map` (of string keys), and when `Map.fetch(map, Atom.to_string(field))` (note diff. 1st arg as above) returns `:error` we ignore it an move on - this behavior makes sense for parsing embeds.

thus, in order to pass map of atoms/strings from `Ecto.load` to `Schema.__load__` I think a lot of changes would need to happen there to keep all of the existing behavior. I'm still not 100% happy with impl., but I couldn't find a cleaner solution yet. Perhaps it's best to sit on it for a few days (e.g. after weekend) :-)

Btw, I took a stab at adding docs and added some more tests. Thanks for the feedback so far, it has been a great learning exp!

wojtekmach commented:
Ok, I've made one final change in `66eb3452b4280bda2cb8890d4805984fca68b8cd`: changes how we treat invalid keys passed to `Ecto.load` (now it ignores instead of raising) which mimics `Kernel.struct` and simplifies the implementation too. I think the're pros and cons to either ignoring or raising, hard to say which one is better.

josevalim commented on lib/ecto/schema.ex:
```diff
@@ -1329,39 +1329,53 @@ defmodule Ecto.Schema do
     end
   end
 
-  defp do_load(struct, types, map, loader) when is_map(map) do
```
Let's add `@doc false`.

josevalim commented on lib/ecto.ex:
```diff
@@ -595,4 +595,51 @@ defmodule Ecto do
+  The first argument can be a schema, or a map (of types) and determines the return value:
+  a struct or a map, respectively.
+
+  The second argument `data` can be a map, a keyword list, or a `{fields, values}` tuple.
```
We should probably mention we allow the keys to be atoms or strings.

josevalim commented:
I have added two final comments and this is good to go! :D

wojtekmach commented:
@josevalim thanks, updated (I changed some wording in the words too)!

btw, here's how it can be used for the original use case of parsing raw query results :)

```elixir
iex> result = Ecto.Adapters.SQL.query!(HexWeb.Repo, "SELECT * FROM users", [])
iex> Enum.map(result.rows, &Ecto.load(HexWeb.User, {result.columns, &1}))
[%HexWeb.User{}, ...]
```

josevalim commented:
Should we include such example in the docs? Also should we mention what
happens for keys that are not in the schema and make sure we have a test
(in case we don't have one yet)?


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

wojtekmach commented:
Yeah, I could mentioned something like this at the end of examples:

```markdown
## Examples

(...)

`Ecto.load/2` is especially useful when parsing raw query results:

    iex> result = Ecto.Adapters.SQL.query!(Repo, "SELECT * FROM users", [])
    iex> Enum.map(result.rows, &Ecto.load(User, {result.columns, &1}))
    [%User{...}, ...]
```

re invalid keys, we have tests and this in docs:

```
Fields that are not present in the schema (or `types` map) are ignored. Values with incorrect types raise an error.
```

josevalim commented:
Sounds good!

On Friday, October 7, 2016, Wojtek Mach <notifications@github.com> wrote:

> Yeah, I could mentioned something like this at the end of examples:
>
> ## Examples
>
> (...)
> `Ecto.load/2` is especially useful when parsing raw query results:
>
>     iex> result = Ecto.Adapters.SQL.query!(Repo, "SELECT * FROM users", [])
>     iex> Enum.map(result.rows, &Ecto.load(User, {result.columns, &1}))
>     [%User{...}, ...]
>
> re invalid keys, we have tests and this in docs:
>
> Fields that are not present in the schema (or `types` map) are ignored. Values with incorrect types raise an error.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-ecto/ecto/pull/1727#issuecomment-252191336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAlbtVGkfJp_kv4dpgcfk2zPos81u5sks5qxg3DgaJpZM4KNAAk>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

wojtekmach commented:
@josevalim @michalmuskala I think this is ready, could you take a final look, esp. at docs & tests? Thanks!

michalmuskala commented:
The only thing I wonder about - is it correct that we use `Ecto.Type.load/2`? Don't we need to do adapter-specific loading in that case? This would require making this a `Ecto.load/3` and accepting a repo as an argument.

To properly test this and see if that's the case, we would need tests in the integration suite for things like booleans, dates and embeds.

josevalim commented:
@michalmuskala this means we would need to move this function to the repository. Maybe we should?

wojtekmach commented:
> @michalmuskala this means we would need to move this function to the repository. Maybe we should?

ping :) If it would be on Repo, how would it look like? `MyApp.Repo.load(schema_or_types, data)`?

josevalim commented:
@wojtekmach I would say so. We should still keep Ecto.load to use without an adapter though.

wojtekmach commented:
@josevalim @michalmuskala I've added a similar `Ecto.Repo.load/2` function and an integration test in 8f47201558ccbd32cecc7a0e04ac79369f55c076

josevalim commented:
@wojtekmach it looks good to me. @michalmuskala, shall we ship it?

wojtekmach commented on lib/ecto.ex:
```diff
@@ -595,4 +595,67 @@ defmodule Ecto do
+  Fields can be atoms or strings.
+
+  Fields that are not present in the schema (or `types` map) are ignored. Values with incorrect
+  types raise an error.
```
thanks, I changed the wording to read: `If any of the values has invalid type, an error is raised.`

michalmuskala commented:
@wojtekmach We talked a bit about this with @josevalim and I came to conclusion that given we now have `Repo.load/2`, I'm not sure I see the use case for `Ecto.load/2`.

* for data coming from database, I'd use `Repo.load/2`
* for external data I'd use changesets
* for programatically generated data that doesn't need processing, I'd use `Kernel.struct/2`

Unless we can find a use case for `Ecto.load/2`, I think we should drop it.

That said I'm really sorry for torturing you with all those changes and going back-and-forth on this so many times. Once again - you're doing an awesome job.

wojtekmach commented:
@michalmuskala @josevalim keeping just Repo.load sounds good to me, it's much easier to add Ecto.load than drop it later. Np about back and forths, I picked up this issue out of the blue and started with a really rough version anyway 😄 I'll ping you when I'm done with changes

wojtekmach commented on test/ecto/repo_test.exs:
```diff
@@ -378,4 +381,60 @@ defmodule Ecto.RepoTest do
+    # # invalid value
+    # assert_raise ArgumentError, "cannot load `0` as type :string for :x in schema Ecto.RepoTest.MySchema", fn ->
+    #   TestRepo.load(MySchema, %{x: 0})
+    # end
```
just double-checking: it looks like a difference between `Ecto.Type.load` and `Ecto.Type.adapter_load` wrt fields of invalid type; the latter allowed loading `0` into a `string` field. Is that ok? If so, I'd need to remove this sentence from docs `If any of the values has invalid type, an error is raised.` or change it slightly.

michalmuskala replied:
The difference is that adapters may customize some types. For example MySQL will be loading `0`, `1`, `<<0>>`, `<<1>>` as booleans (since it lacks native boolean types). There's also special handling for json - mysql will load json-encoded strings and all SQL adapters will have special handling for loading embeds. Furthermore, every adapter must specify how to load the special `binary_id` type that is adapter-dependent.

The lack of error here is strange... It should be wrong.

@josevalim Is that because of this bypass? https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/type.ex#L631

josevalim commented:
I am fixing the `@bypass` thing on master.

josevalim commented:
`@bypass` has been removed from master. please rebase and you should be good to go!

wojtekmach commented:
@josevalim thanks for a quick fix, I rebased and re-enabled that commented out test

josevalim commented:
Beautiful! Let us know when this is ready to go! :D

wojtekmach commented:
from my end I think it's ready. FYI: couple commits ago I did small changes to unit tests (removed some examples) and docs, but other than that I think it's mostly unchanged. Let me know if anything else is missing!

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

wojtekmach commented:
@josevalim @michalmuskala I was just using this function in iex after a break and it got me thinking. Per these use cases:

> * for data coming from database, I'd use Repo.load/2
> * for external data I'd use changesets
> * for programatically generated data that doesn't need processing, I'd use Kernel.struct/2

if the use case for this function is really just loading data from DB, where usually it would come as columns & rows, perhaps it would make sense to make it more convenient. Instead of:

```elixir
    iex> result = MyRepo.query!("SELECT * FROM users", [])
    iex> Enum.map(result.rows, &MyRepo.load(User, {result.columns, &1}))
    [%User{}, ...]
```

we could just do:

```elixir
    iex> result = MyRepo.query!("SELECT * FROM users", []) ; MyRepo.load(User, result.columns, result.rows) # or {result.columns, result.rows}
    [%User{}, ...]
```

this could replace the form `load(User, {columns, row})`. There's nothing wrong with doing the `Enum.map(result.rows, &Repo.load(...))` of course, but I'm just thinking about optimizing for the most important use case. Keeping it as is if fine too!

Also, since it was a really late addition to rc.3 perhaps it would be better to push it to v2.2.0 if it needs  any more discussion. Sorry to keep bugging you about again! :)

josevalim commented:
@wojtekmach the problem is that not all adapters return data in that format. :(

wojtekmach commented:
yeah, I figured mongodb would return it as json or something. In that case, isn't that an argument _against_  `Repo.load(User, {columns, row})` which we already added here? Anyhow, you guys probably know better so it's probably better to leave it as is - sorry for the hassle and thanks for the merge :smile:

josevalim commented:
> sorry for the hassle

This is no problem at all. This is a tricky decision because we are going with a more low-level construct. Let's see how it goes. We can always add Repo.load_all in the future.
